### PR TITLE
[FLINK-29271]Change to byte array from bytebuffer to improve performance and compatible direct byte buffers

### PR DESCRIPTION
**Pull request by jiangjiguang**
## What is the purpose of the change
The PR uses byte array api instead of ByteBuffer api to improve performance  and compatible with DirectByteBuffer which does not have byte array backed  when reading parquet DataPage V2

## Brief change log
  - Use the byte array api instead of ByteBuffer api.

## Verifying this change
I have done a test to compare byte array api and ByteBuffer api on parquet-mr project.
The result is:
[Unpack8ValuesByteArray spent time] 80 ms
[Unpack8ValuesByteBuffer spent time] 133 ms

The test code is:
package org.apache.parquet.column.values.bitpacking;

import java.nio.ByteBuffer;

public class ByteBufferTest {
  private static final BytePacker bytePacker = Packer.LITTLE_ENDIAN.newBytePacker(7);

  private static final int COUNT = 100000;

  public static void main(String[] args) {
    byte  [] in  = new byte[1008];
    int [] out = new int[1152];
    int [] out1 = new int[1152];
    int [] out2 = new int[1152];

    int res = 0;

    for(int i = 0; i < in.length; i++) {
      in[i] = (byte) i;
    }

    for(int i = 0; i < COUNT; i++) {
      res += unpack8ValuesBytes(in, out, i % out.length);
    }

    res = 0;
    long t1 = System.currentTimeMillis();
    for(int i = 0; i < COUNT; i++) {
      res += unpack8ValuesBytes(in, out1, i % out.length);
    }
    long t2 = System.currentTimeMillis();
    System.out.println("[Unpack8ValuesByteArray spent time] " + (t2-t1) + " ms");

    ByteBuffer byteBuffer = ByteBuffer.wrap(in);

    for(int i = 0; i < COUNT; i++) {
      res += unpack8ValuesByteBuffer(byteBuffer, out, i % out.length);
    }

    res = 0;
    long t3 = System.currentTimeMillis();
    for(int i = 0; i < COUNT; i++) {
      res += unpack8ValuesByteBuffer(byteBuffer, out2, i % out.length);
    }
    long t4 = System.currentTimeMillis();
    System.out.println("[Unpack8ValuesByteBuffer spent time] " + (t4-t3) + " ms");

    for (int i=0; i<out1.length; i++) {
      if(out1[i] != out2[i]) {
        System.out.println("diff: " + out1[i] + " " + out2[i]);
      }
    }
  }

  private static int unpack8ValuesBytes(byte [] in, int [] out, int ctr) {
    for(int i = 0, j = 0; i < in.length; i+=7, j+=8) {
      bytePacker.unpack8Values(in, i, out, j);
    }
    return out[ctr];
  }
  private static int unpack8ValuesByteBuffer(ByteBuffer in, int [] out, int ctr) {
    for(int i = 0, j = 0; i < in.capacity(); i+=7, j+=8) {
      bytePacker.unpack8Values(in, i, out, j);
    }
    return out[ctr];
  }
}

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? (yes / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by jiangjiguang** [comment]
@flinkbot run azure re-run the last Azure build

**Comment by wuchong** [comment]
@lsyldliu do you want to review this?

**Comment by lsyldliu** [comment]
@jiangjiguang Thanks for contribution, I will review it as soon as posssible

**Comment by jiangjiguang** [comment]
@lsyldliu Is there a progress ?

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/RunLengthDecoder.java [hunk 1]
```diff
@@ -265,8 +265,16 @@ void readNextGroup() {
                     while (valueIndex < this.currentCount) {
                         // values are bit packed 8 at a time, so reading bitWidth will always work
                         ByteBuffer buffer = in.slice(bitWidth);
-                        this.packer.unpack8Values(
-                                buffer, buffer.position(), this.currentBuffer, valueIndex);
+                        if (buffer.hasArray()) {
+                            this.packer.unpack8Values(
```
Comment (luoyuxia) [inline flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/RunLengthDecoder.java:269]: I'm wondering why this changes will improve perfermance. IIUC, from the code
```java
public void unpack8Values(final byte[] input, final int inPos, final int[] output, final int outPos) {
    unpack8Values(ByteBuffer.wrap(input), inPos, output, outPos);
  }
```
 the method just delegates to `unpack8Values` and it needs extra `ByteBuffer.wrap(input)`.
Could you please explain a little bit?

**Review by luoyuxia — COMMENTED** [review]
@jiangjiguang Thanks for contribution. I left some comments. PTAL.

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/RunLengthDecoder.java (hunk 1 above)
Comment (jiangjiguang) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/RunLengthDecoder.java:269]: @luoyuxia This BytePacker is only abstract parent class, and unpack8Values will not run this method.
I have debug code, it run Packer7 class which is code generated.

**Comment by jiangjiguang** [comment]
image

**Comment by jiangjiguang** [comment]
image

**Comment by jiangjiguang** [comment]
@luoyuxia  packer is made by  this.packer = Packer.LITTLE_ENDIAN.newBytePacker(bitWidth); so it impossible to run the  abstract parent class BytePacker'method

image

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/RunLengthDecoder.java (hunk 1 above)
Comment (luoyuxia) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/RunLengthDecoder.java:269]: Thanks for explanation. Make sense to me.

Comment (luoyuxia) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/RunLengthDecoder.java:269]: Could you please add some comments in here to expain why we change it?
Now, we use a deprecated method, we may need some comments for it so that the others can know the context when come to this method for the case that the method is deleted by a higher parquet version in some days and flink needs to upgrad to the  higher parquet version.

**Comment by luoyuxia** [comment]
@jiangjiguang Also, could you please help rebase the master?

File: flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/RunLengthDecoder.java (hunk 1 above)
Comment (jiangjiguang) [reply flink-formats/flink-parquet/src/main/java/org/apache/flink/formats/parquet/vector/reader/RunLengthDecoder.java:270]: @luoyuxia Good idea. I have added comments.  Besides, I have connected parquet-mr community to suggest remote the deprecated flag and change to use byte array default.

**Comment by jiangjiguang** [comment]
@luoyuxia I have rebased the master

**Review by luoyuxia — APPROVED** [review]
Thanks for updating. LGTM

**Comment by jiangjiguang** [comment]
@wuchong Could you please help merge this pr, thanks~
